sgl-kernel/csrc/cpu/gemm_fp8.cpp

```diff
@@ -0,0 +1,356 @@
+
+        __m256i b8_0 = _mm512_extracti32x8_epi32(b8, 0);
+        __m256i b8_1 = _mm512_extracti32x8_epi32(b8, 1);
+
+        __m512bh bf16_0 = CVT_FP8_TO_BF16(b8_0);
+        __m512bh bf16_1 = CVT_FP8_TO_BF16(b8_1);
+
+        // Apply scale
+        __m512 f0_lo = CVT_BF16_TO_FP32(_mm512_extracti32x8_epi32((__m512i)bf16_0, 0));
+        __m512 f0_hi = CVT_BF16_TO_FP32(_mm512_extracti32x8_epi32((__m512i)bf16_0, 1));
+        __m512 f1_lo = CVT_BF16_TO_FP32(_mm512_extracti32x8_epi32((__m512i)bf16_1, 0));
+        __m512 f1_hi = CVT_BF16_TO_FP32(_mm512_extracti32x8_epi32((__m512i)bf16_1, 1));
+
+        f0_lo = _mm512_mul_ps(f0_lo, vd);
+        f0_hi = _mm512_mul_ps(f0_hi, vd);
+        f1_lo = _mm512_mul_ps(f1_lo, vd);
+        f1_hi = _mm512_mul_ps(f1_hi, vd);
+
+        bf16_0 = _mm512_cvtne2ps_pbh(f0_hi, f0_lo);
+        bf16_1 = _mm512_cvtne2ps_pbh(f1_hi, f1_lo);
+
+        _mm512_storeu_si512(Btmp + k * ldb_tmp * 2 + n * 2 + 0, (__m512i)bf16_0);
+        _mm512_storeu_si512(Btmp + k * ldb_tmp * 2 + n * 2 + 32, (__m512i)bf16_1);
+    }
+  }
+}
+
+template <typename scalar_t, typename packed_t, bool has_bias>
+struct brgemm {};
+
+template <typename scalar_t, bool has_bias>
+struct brgemm<scalar_t, scalar_t, has_bias> {
+  static inline void apply(
+      const scalar_t* __restrict__ A,
+      const scalar_t* __restrict__ B,
+      scalar_t* __restrict__ C,
+      scalar_t* __restrict__ Btmp,
+      float* __restrict__ Ctmp,
+      const float* __restrict__ bias,
+      const float* __restrict__ scale,
+      int M,
+      int N,
+      int K,
+      int lda,
+      int ldb,
+      int ldc,
+      int64_t blocks_k_per_group) {
+    UNUSED(scale);
+
+    constexpr int BLOCK_N = block_size_n();
+    at::native::cpublas::brgemm(
+        M, N, K, lda, ldb, BLOCK_N, /* add_C */ false, A, B, Ctmp);
+
+    // copy from Ctmp to C
+    for (int m = 0; m < M; ++m) {
+      if constexpr (has_bias) {
+        copy_add_stub(C + m * ldc, Ctmp + m * BLOCK_N, bias, N);
+      } else {
+        copy_stub(C + m * ldc, Ctmp + m * BLOCK_N, N);
+      }
+    }
+  }
+};
+
+template <bool has_bias>
+struct brgemm<at::BFloat16, at::Float8_e4m3fn, has_bias> {
+  static inline void apply(
+      const at::BFloat16* __restrict__ A,
+      const at::Float8_e4m3fn* __restrict__ B,
+      at::BFloat16* __restrict__ C,
+      at::BFloat16* __restrict__ Btmp,
+      float* __restrict__ Ctmp,
+      const float* __restrict__ bias,
+      const float* __restrict__ scale,
+      int M,
+      int N,
+      int K,
+      int lda,
+      int ldb,
+      int ldc,
+      int64_t blocks_k_per_group) {
+    constexpr int BLOCK_N = block_size_n();
+
+    // [BLOCK_K, BLOCK_N] -> [BLOCK_K / 2, BLOCK_N * 2]
+    const int ldb_tmp = block_size_n();
+
+    // accumulate across K per BLOCK_K
+    for (int k = 0; k < K; k += BLOCK_K) {
+      int kb_size = std::min(BLOCK_K, K - k);
+
+      int idx = (k / BLOCK_K) / blocks_k_per_group;
```

Comment:
```suggestion
      int idx = k / BLOCK_K;
```

if we force block_size_K == BLOCK_K, then we can use idx = k / BLOCK_K, however i prefer something like:
```
static_assert(BLOCK_K == 128);

int idx = k >> 7; // this get rid of integer div which might not be optimized by gcc compiler.
```